Codes fail on round trip marshal/unmarshal

Unmarshalling a `codes.Code` value that was previously marshalled to JSON fails.

Test case reproducing the issue:

```
func TestRoundTripCodes(t *testing.T) {
	tests := []struct {
		input  codes.Code
	}{
		{codes.Unset},
		{codes.Error},
		{codes.Ok},
	}
	for _, test := range tests {
		c := test.input
		out := new(codes.Code)

		b, err := c.MarshalJSON()
		if err != nil {
			t.Fatalf("Code(%s).MarshalJSON() errored: %v", test.input, err)
		}

		if err := out.UnmarshalJSON(b); err != nil {
			t.Fatalf("Code.UnmarshalJSON(%q) errored: %v", c, err)
		}

		if *out != test.input {
			t.Errorf("failed to round trip %q, output was %v", test.input, out)
		}
	}
}
```

Fails with (all codes fail, not just Unset):

```
=== RUN   TestRoundTripCodes
    TestRoundTripCodes: roundtrip_codes_test.go:27: Code.UnmarshalJSON("Unset") errored: invalid code: "\"Unset\""
--- FAIL: TestRoundTripCodes (0.00s)
```

Unmarshal/marshal should round trip successfully, but the wrapping quotes in the incoming bytes when performing an unmarshal aren't expected by the mapping code, so it fails.

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -34,7 +34,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - The function signature of the Span `RecordError` method in `go.opentelemetry.io/otel` is updated to no longer take an unused context and instead take a required error value and a variable number of `EventOption`s. (#1254)
 - Move the `go.opentelemetry.io/otel/api/global` package to `go.opentelemetry.io/otel/global`. (#1262)
 - Rename correlation context header from `"otcorrelations"` to `"baggage"` to match the OpenTelemetry specification. (#1267)
-- Remove quotes from `codes.Code` byte representation in `UnmarshalJSON`. (#1276)
+- Fix `Code.UnmarshalJSON` to work with valid json only. (#1276)
 
 ### Removed
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -34,6 +34,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - The function signature of the Span `RecordError` method in `go.opentelemetry.io/otel` is updated to no longer take an unused context and instead take a required error value and a variable number of `EventOption`s. (#1254)
 - Move the `go.opentelemetry.io/otel/api/global` package to `go.opentelemetry.io/otel/global`. (#1262)
 - Rename correlation context header from `"otcorrelations"` to `"baggage"` to match the OpenTelemetry specification. (#1267)
+- Remove quotes from `codes.Code` byte representation in `UnmarshalJSON`. (#1273)
 
 ### Removed
 

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -34,7 +34,7 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 - The function signature of the Span `RecordError` method in `go.opentelemetry.io/otel` is updated to no longer take an unused context and instead take a required error value and a variable number of `EventOption`s. (#1254)
 - Move the `go.opentelemetry.io/otel/api/global` package to `go.opentelemetry.io/otel/global`. (#1262)
 - Rename correlation context header from `"otcorrelations"` to `"baggage"` to match the OpenTelemetry specification. (#1267)
-- Remove quotes from `codes.Code` byte representation in `UnmarshalJSON`. (#1273)
+- Remove quotes from `codes.Code` byte representation in `UnmarshalJSON`. (#1276)
 
 ### Removed
 

--- a/codes/codes.go
+++ b/codes/codes.go
@@ -20,7 +20,6 @@ package codes // import "go.opentelemetry.io/otel/codes"
 
 import (
 	"encoding/json"
-	"errors"
 	"fmt"
 	"strconv"
 )
@@ -75,9 +74,6 @@ func (c *Code) UnmarshalJSON(b []byte) error {
 	var x interface{}
 	if err := json.Unmarshal(b, &x); err != nil {
 		return err
-	}
-	if x == nil {
-		return errors.New("null code value")
 	}
 	switch x.(type) {
 	case string:

--- a/codes/codes.go
+++ b/codes/codes.go
@@ -19,6 +19,7 @@
 package codes // import "go.opentelemetry.io/otel/codes"
 
 import (
+	"bytes"
 	"fmt"
 	"strconv"
 )
@@ -70,6 +71,7 @@ func (c *Code) UnmarshalJSON(b []byte) error {
 		return fmt.Errorf("nil receiver passed to UnmarshalJSON")
 	}
 
+	b = bytes.Replace(b, []byte("\""), []byte(""), -1)
 	if ci, err := strconv.ParseUint(string(b), 10, 32); err == nil {
 		if ci >= maxCode {
 			return fmt.Errorf("invalid code: %q", ci)

--- a/codes/codes.go
+++ b/codes/codes.go
@@ -19,7 +19,8 @@
 package codes // import "go.opentelemetry.io/otel/codes"
 
 import (
-	"bytes"
+	"encoding/json"
+	"errors"
 	"fmt"
 	"strconv"
 )
@@ -46,9 +47,9 @@ var codeToStr = map[Code]string{
 }
 
 var strToCode = map[string]Code{
-	"Unset": Unset,
-	"Error": Error,
-	"Ok":    Ok,
+	`"Unset"`: Unset,
+	`"Error"`: Error,
+	`"Ok"`:    Ok,
 }
 
 // String returns the Code as a string.
@@ -71,21 +72,33 @@ func (c *Code) UnmarshalJSON(b []byte) error {
 		return fmt.Errorf("nil receiver passed to UnmarshalJSON")
 	}
 
-	b = bytes.Replace(b, []byte("\""), []byte(""), -1)
-	if ci, err := strconv.ParseUint(string(b), 10, 32); err == nil {
-		if ci >= maxCode {
-			return fmt.Errorf("invalid code: %q", ci)
+	var x interface{}
+	if err := json.Unmarshal(b, &x); err != nil {
+		return err
+	}
+	if x == nil {
+		return errors.New("null code value")
+	}
+	switch x.(type) {
+	case string:
+		if jc, ok := strToCode[string(b)]; ok {
+			*c = jc
+			return nil
 		}
+		return fmt.Errorf("invalid code: %q", string(b))
+	case float64:
+		if ci, err := strconv.ParseUint(string(b), 10, 32); err == nil {
+			if ci >= maxCode {
+				return fmt.Errorf("invalid code: %q", ci)
+			}
 
-		*c = Code(ci)
-		return nil
+			*c = Code(ci)
+			return nil
+		}
+		return fmt.Errorf("invalid code: %q", string(b))
+	default:
+		return fmt.Errorf("invalid code: %q", string(b))
 	}
-
-	if jc, ok := strToCode[string(b)]; ok {
-		*c = jc
-		return nil
-	}
-	return fmt.Errorf("invalid code: %q", string(b))
 }
 
 // MarshalJSON returns c as the JSON encoding of c.

--- a/codes/codes_test.go
+++ b/codes/codes_test.go
@@ -16,6 +16,7 @@ package codes
 
 import (
 	"bytes"
+	"encoding/json"
 	"fmt"
 	"testing"
 )
@@ -61,17 +62,18 @@ func TestCodeUnmarshalJSON(t *testing.T) {
 		want  Code
 	}{
 		{"0", Unset},
-		{"Unset", Unset},
+		{`"Unset"`, Unset},
 		{"1", Error},
-		{"Error", Error},
+		{`"Error"`, Error},
 		{"2", Ok},
-		{"Ok", Ok},
+		{`"Ok"`, Ok},
 	}
 	for _, test := range tests {
 		c := new(Code)
 		*c = Code(maxCode)
-		if err := c.UnmarshalJSON([]byte(test.input)); err != nil {
-			t.Fatalf("Code.UnmarshalJSON(%q) errored: %v", test.input, err)
+
+		if err := json.Unmarshal([]byte(test.input), c); err != nil {
+			t.Fatalf("json.Unmarshal(%q, Code) errored: %v", test.input, err)
 		}
 		if *c != test.want {
 			t.Errorf("failed to unmarshal %q as %v", test.input, test.want)
@@ -83,11 +85,13 @@ func TestCodeUnmarshalJSONErrorInvalidData(t *testing.T) {
 	tests := []string{
 		fmt.Sprintf("%d", maxCode),
 		"Not a code",
+		"Unset",
+		"true",
 	}
 	c := new(Code)
 	for _, test := range tests {
-		if err := c.UnmarshalJSON([]byte(test)); err == nil {
-			t.Fatalf("Code.UnmarshalJSON(%q) did not error", test)
+		if err := json.Unmarshal([]byte(test), c); err == nil {
+			t.Fatalf("json.Unmarshal(%q, Code) did not error", test)
 		}
 	}
 }

--- a/codes/codes_test.go
+++ b/codes/codes_test.go
@@ -87,6 +87,8 @@ func TestCodeUnmarshalJSONErrorInvalidData(t *testing.T) {
 		"Not a code",
 		"Unset",
 		"true",
+		`"Not existing"`,
+		"",
 	}
 	c := new(Code)
 	for _, test := range tests {

--- a/codes/codes_test.go
+++ b/codes/codes_test.go
@@ -133,3 +133,30 @@ func TestCodeMarshalJSONErrorInvalid(t *testing.T) {
 		t.Fatal("Code(maxCode).MarshalJSON() returned non-nil value")
 	}
 }
+
+func TestRoundTripCodes(t *testing.T) {
+	tests := []struct {
+		input Code
+	}{
+		{Unset},
+		{Error},
+		{Ok},
+	}
+	for _, test := range tests {
+		c := test.input
+		out := new(Code)
+
+		b, err := c.MarshalJSON()
+		if err != nil {
+			t.Fatalf("Code(%s).MarshalJSON() errored: %v", test.input, err)
+		}
+
+		if err := out.UnmarshalJSON(b); err != nil {
+			t.Fatalf("Code.UnmarshalJSON(%q) errored: %v", c, err)
+		}
+
+		if *out != test.input {
+			t.Errorf("failed to round trip %q, output was %v", test.input, out)
+		}
+	}
+}

